final_task/rss_reader/articles_handler.py

```diff
@@ -0,0 +1,226 @@
+import dataclasses
+import datetime
+import html
+import json
+import logging
+import os
+import urllib.request
+from typing import List, Tuple, Optional
+
+import dateutil.parser
+from bs4 import BeautifulSoup
+import feedparser
+
+import custom_error
+import single_article
+
+directory_to_module = os.path.abspath(os.path.dirname(__file__))
+CACHE_FILE_NAME = os.path.join(directory_to_module, 'cache.json')
+
+
+def parse_rss(rss_url: str) -> feedparser.FeedParserDict:
+    """Parse the rss"""
+    return feedparser.parse(rss_url)
+
+
+def create_rss_json(data: List[single_article.SingleArticle]):
+    """Prints result as JSON in stdout"""
+    return json.dumps(data, cls=single_article.EnhancedJSONEncoder, ensure_ascii=False)
+
+
+def print_article(data: single_article.SingleArticle) -> None:
+    """Prints a single article in stdout"""
+    print("-" * 120)
+    print(data)
+    print("-" * 120)
+
+
+def print_rss_articles(rss_articles: List[single_article.SingleArticle]) -> None:
+    """Prints source feed and articles"""
+    for article in rss_articles:
+        print_article(article)
+
+
+def find_links_in_article(article: feedparser.FeedParserDict) -> Tuple[List, str]:
+    """Finds images and links in the article"""
+    all_links = []
+
+    soup = BeautifulSoup(article['summary'], features="html.parser")
+    for index, img in enumerate(soup.find_all('img'), 1):
+        if (str(img['src']) == '') and (str(img['alt']) == ''):
+            continue
+        if str(img['alt']) == '':
+            all_links.append([str(img['src']), 'image'])
+            img.replaceWith(f"[image {index}: no description][{index}] ")
+        elif str(img['src']) == '':
+            img.replaceWith(f"[image(no link): {img['alt']}] ")
+        else:
+            all_links.append([str(img['src']), 'image'])
+            img.replaceWith(f"[image {index}: {img['alt']}][{index}] ")
+
+    for link in article['links']:
+        if link['href'] not in all_links:
+            all_links.append([link['href'], 'other'])
+            """if link['href'] == article['link']:
+                all_links.append([link['href'], 'text'])
+            else:
+                head = requests.head(link['href'])
+                header = head.headers
+                content_type = header.get('content-type')
+
+                if 'image' in content_type:
+                    all_links.append([link['href'], 'image'])
+                else:
+                    all_links.append([link['href'], 'text'])"""
+
+    return all_links, str(soup.text)
+
+
+def unescape(text_to_unescape: str) -> str:
+    """Unescape text"""
+    return html.unescape(text_to_unescape)
+
+
+def is_connected():
+    hostname = 'https://www.google.com/'
+    try:
+        urllib.request.urlopen(hostname)
+        return True
+    except urllib.request.HTTPError:
+        return False
+    except urllib.request.URLError:
```

Comment:
1. How can you have URLError with this static link?
2. Does URLError occurrence says that you have no internet connection?